src/connect4/ai.rs

```diff
@@ -2,196 +2,252 @@
-        self.board.add_disc(moveCol, team);
-    }
-
-    fn compare(&self, other: MoveCheck) -> i32 {
-        for i in (0..4).rev() {
-            if self.runs[i] != other.runs[i] {
-                return i*(self.runs[i]-other.runs[i]);
-            }
-        }
-        0
-    }
-
-    fn has_end_result(&self) -> bool {
-        self.runs[4] > 0
-    }
 }
 
+//Ordering implementation based on documentation example (https://doc.rust-lang.org/std/cmp/trait.Ord.html), tailored to compare MoveCheck's runs
 impl Ord for MoveCheck {
     fn cmp(&self, other: &Self) -> Ordering {
         for i in (0..4).rev() {
             if self.runs[i] != other.runs[i] {
-                if i*(self.runs[i]-other.runs[i]) < 0 {
-                    Ordering::Less
+                if (i as i32)*(self.runs[i]-other.runs[i]) < 0 {
+                    return Ordering::Less;
                 } else {
-                    Ordering::Greater
+                    return Ordering::Greater;
                 }
             }
         }
         Ordering::Equal
     }
 }
 
-/*impl PartialOrd for MoveCheck {
+impl PartialOrd for MoveCheck {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
+
 impl PartialEq for MoveCheck {
     fn eq(&self, other: &Self) -> bool {
-        self.height == other.height
+        self.runs == other.runs
     }
-}*/
+}
+
+impl Eq for MoveCheck {}
 
-struct AI {
+pub struct AI {
     team: i32,
     difficulty: i32
 }
 
 impl AI {
-    fn pick_optimal_move(state: GameState) -> i32 {
-        let mut startBoard = state.get_board().clone();
-        0
+    fn new(team: i32, difficulty: i32) -> Self {
+        AI { team, difficulty }
+    }
+
+    fn pick_optimal_move(&self, state: GameState) -> i32 {
+        let mut bestMove = -1;
+        let mut bestProb = 0.0;
+        for i in 0..BOARD_SIZE.1 {
+            let mut currBoard = state.board.clone();
+            currBoard.insert(i, self.team, MyColor::White);
+            let currProb = self.find_win_probability(currBoard, 0, self.difficulty);
+            if currProb == 1.0 {
+                return i as i32;
+            } else if currProb >= bestProb {
+                bestProb = currProb;
+                bestMove = i;
+            }
+        }
+        bestMove
     }
 
-    fn find_win_probability(&self, board: GameState, currMove: i32, lastMove: i32) -> f32 {
+    fn find_win_probability(&self, board: Board, currMove: i32, lastMove: i32) -> f32 {
         //Check win for AI turn
-        let moves = Vec::new();
-        for i in 0..core::BOARD_SIZE.0 {
-            if currMove & 2 == 0 {
-                let oppMove = MoveCheck::init(board, i, self.team%2 + 1);
-                //If move would cause other team to win, return 0 as prob. Otherwise, call recursively on next move
-                if oppMove.has_end_result() {
-                    return 0;
+        let mut moves = Vec::new();
+        let mut out = "".to_owned();
```

Comment:
This is in for debugging I was doing. I would remove, but I'm likely going to need it again when I look at this more thoroughly.